src/main/scala/com/advancedtelematic/director/manifest/DeviceManifestUpdate.scala

```diff
@@ -1,28 +1,47 @@
 package com.advancedtelematic.director.manifest
 
+import java.security.MessageDigest
+
+import akka.http.scaladsl.util.FastFuture
 import cats.syntax.show._
 import com.advancedtelematic.director.data.Codecs._
 import com.advancedtelematic.director.data.DeviceRequest.{CustomManifest, EcuManifest}
-import com.advancedtelematic.director.db.{DeviceRepositorySupport, DeviceUpdate, DeviceUpdateResult, UpdateTypesRepositorySupport}
+import com.advancedtelematic.director.db.{DeviceRepositorySupport, DeviceUpdate, DeviceUpdateResult, ProcessedManifestsRepositorySupport, UpdateTypesRepositorySupport}
 import com.advancedtelematic.director.manifest.Verifier.Verifier
 import com.advancedtelematic.libats.data.DataType.Namespace
 import com.advancedtelematic.libats.messaging_datatype.DataType.{DeviceId, EcuSerial}
-import com.advancedtelematic.libtuf.data.TufDataType.{JsonSignedPayload, OperationResult, SignedPayload, TufKey}
+import com.advancedtelematic.libtuf.data.TufDataType.{OperationResult, SignedPayload, TufKey}
 import io.circe.Json
 import org.slf4j.LoggerFactory
 
 import scala.async.Async._
 import scala.concurrent.{ExecutionContext, Future}
 import slick.driver.MySQLDriver.api._
+import com.advancedtelematic.libtuf.crypt.CanonicalJson._
 
 class DeviceManifestUpdate(afterUpdate: AfterDeviceManifestUpdate,
                            verifier: TufKey => Verifier
                           )(implicit val db: Database, val ec: ExecutionContext)
-    extends DeviceRepositorySupport
-    with UpdateTypesRepositorySupport {
+    extends DeviceRepositorySupport with UpdateTypesRepositorySupport with ProcessedManifestsRepositorySupport {
   private lazy val _log = LoggerFactory.getLogger(this.getClass)
+  val digester: MessageDigest = MessageDigest.getInstance("SHA-256")
+
+  def setDeviceManifest(namespace: Namespace, device: DeviceId, signedDevMan: SignedPayload[Json]): Future[Unit] = {
+    val digest = digester.digest(signedDevMan.signed.canonical.getBytes("UTF-8"))
+    val hex = digest.map("%02x".format(_)).mkString
+
+    processedManifestsRepository.contains(namespace, device, hex).flatMap { alreadyProcessed =>
+      if (alreadyProcessed) {
+        FastFuture.successful(())
+      } else {
+        processDeviceManifest(namespace, device, signedDevMan).flatMap { _ =>
+          processedManifestsRepository.add(namespace, device, hex).map(_ => ())
+        }
+      }
+    }
+  }
```

Comment:
Could we somehow incorporate this check with the current states? Because currently there is already `NoChange` [1], isn't this an extension of that? 

[1]: https://github.com/advancedtelematic/director/blob/53de30fb27bfefd4c7a0c9bfeb0ab35193ff3a01/src/main/scala/com/advancedtelematic/director/manifest/AfterUpdate.scala#L44

Reply:
That's a good idea I didn't consider, but my mental model is more that of an "incoming cache". I find the code in `DeviceUpdate` already hard to follow and would think adding another `DeviceUpdateResult` would make it even worse.

Reply:
The code in `DeviceManifestUpdate` is simpler I guess.